### PR TITLE
frontend: Use an uncommon name for csrf cookie

### PR DESCRIPTION
**Pull request by keegancsmith**
The name `csrf_token` seems to be use by other applications. When a user views
Sourcegraph on localhost it is not uncommon for a csrf_token to be set by
another app the user uses on localhost. This has lead to csrf errors, making the
site unuseable for the user. The proposed solution is to just use a different
name.

This does mean when this rolls out users will not have `sg_csrf_token` set. I
suspect they will run into a not authorized error, but will work on the next
page refresh.

Fixes https://github.com/sourcegraph/sourcegraph/issues/65

### CODE REVIEW

**Comment by kevinburke** [comment]
Shouldn't there be a corresponding change to read the cookie from the new name?

**Comment by nicksnyder** [comment]
I believe the protect method here does both functions (verify and set)

**Comment by kevinburke** [comment]
oh ok!

**Comment by sqs** [comment]
Good call! I am trying to understand the impact of this.

> This does mean when this rolls out users will not have sg_csrf_token set. I suspect they will run into a not authorized error, but will work on the next page refresh.

If you have a Sourcegraph tab open from BEFORE the rename, then the Sourcegraph endpoint is upgraded to incorporate this rename, your requests for certain API endpoints would fail.

But this affects very few or maybe no actual endpoints because most of our endpoints are protected by CORS and not CSRF tokens. (I [don't see any `r.Form.Get` etc.](https://sourcegraph.com/search?q=repo:sourcegraph/sourcegraph%24+%5C.Form%5C.))

What are the operations that would fail (until a page refresh)?

**Comment by slimsag** [comment]
I am OK with this change but only if we prove out what exactly the behavior will be for existing users that have a `csrf_token` as Quinn mentioned above.

Also, there is a more general internal issue here around the question of what our CSRF tokens protect today and whether or not removing them outright is a good idea: https://github.com/sourcegraph/enterprise/issues/11121

**Comment by keegancsmith** [comment]
TL;DR Rare POST requests will fail until reload

Yeah, I'm a bit confused. So this is what happens when the CSRF cookie doesn't exist:

- `ErrNoCookie` is returned when we get the token https://github.com/sourcegraph/sourcegraph/blob/3d45734524ae49c407ea7639dbb6d0ba96e6a59d/vendor/github.com/gorilla/csrf/csrf.go#L193
- We then set the cookie to a new generated and correct token https://github.com/sourcegraph/sourcegraph/blob/3d45734524ae49c407ea7639dbb6d0ba96e6a59d/vendor/github.com/gorilla/csrf/csrf.go#L194-L213
- If we are not a safe method we then check the request token https://github.com/sourcegraph/sourcegraph/blob/3d45734524ae49c407ea7639dbb6d0ba96e6a59d/vendor/github.com/gorilla/csrf/csrf.go#L194-L213 https://github.com/sourcegraph/sourcegraph/blob/3d45734524ae49c407ea7639dbb6d0ba96e6a59d/vendor/github.com/gorilla/csrf/csrf.go#L254-L259

The important part here is "unsafe". So where do we actually use the CSRF middleware? Just the app routes: [References](https://sourcegraph.sgdev.org/github.com/sourcegraph/sourcegraph@3d45734524ae49c407ea7639dbb6d0ba96e6a59d/-/blob/cmd/frontend/internal/pkg/handlerutil/middleware.go#L12:6&tab=references); [POST routes](https://sourcegraph.sgdev.org/search?q=repo:%5Egithub%5C.com/sourcegraph/sourcegraph%24%403d45734+file:%5Ecmd/frontend/internal/app/router/+%22POST%22)
image

https://github.com/sourcegraph/sourcegraph/blob/3d45734/cmd/frontend/internal/app/router/router.go#L54-L98

So if you do one of the POST requests above it will fail until you reload the page. Note that all those POST requests are rare! The chances of someone loading the sign-in/sign-up/password-reset page; we deploy this change; they submit the form is likely _very_ rare.

**Comment by keegancsmith** [comment]
ping @slimsag @nicksnyder
